src/streams.rs

```diff
@@ -107,16 +109,20 @@ impl BaseStream {
     }
 
     fn connect_tcp(host: &Host<&str>, port: u16, info: &ConnectInfo) -> Result<(TcpStream, Option<mpsc::Sender<()>>)> {
-        let stream = happy::connect(host, port, info.base_settings.connect_timeout)?;
+        let stream = happy::connect(host, port, info.base_settings.connect_timeout, info.deadline)?;
         stream.set_read_timeout(Some(info.base_settings.read_timeout))?;
         let timeout = info
-            .base_settings
-            .timeout
-            .map(|timeout| -> Result<mpsc::Sender<()>> {
+            .deadline
+            .map(|deadline| -> Result<mpsc::Sender<()>> {
                 let stream = stream.try_clone()?;
                 let (tx, rx) = mpsc::channel();
                 thread::spawn(move || {
-                    if let Err(mpsc::RecvTimeoutError::Timeout) = rx.recv_timeout(timeout) {
+                    let shutdown = match deadline.checked_duration_since(Instant::now()) {
+                        Some(timeout) => rx.recv_timeout(timeout) == Err(mpsc::RecvTimeoutError::Timeout),
+                        None => rx.try_recv() == Err(mpsc::TryRecvError::Empty),
```

Comment:
I'm trying to understand this line. If the deadline is expired and the channel is still connected, shutdown the stream, right?

Reply:
Exactly, the other error is `mpsc::TryRecvError::Disconnected` which means the request is already done when we reach this line which might happen if the thread is scheduled in after a significant delay. (`Ok(_)` should never happen as we use only disconnection to signal this.)

Reply:
Sounds good